api/src/main/java/org/pmiops/workbench/cohortbuilder/querybuilder/MeasurementQueryBuilder.java

```diff
@@ -0,0 +1,163 @@
+    Map<String, QueryParameterValue> queryParams = new HashMap<>();
+    for (SearchParameter parameter : parameters.getParameters()) {
+      validateAttributes(parameter);
+      if (parameter.getConceptId() == null) {
+        throw new BadRequestException("Please provide valid concept id for Measurements.");
+      }
+      String baseSql = MEASUREMENT_SQL_TEMPLATE.replace("${conceptId}", parameter.getConceptId().toString());
+      List<String> tempQueryParts = new ArrayList<String>();
+      for (Attribute attribute : parameter.getAttributes()) {
+        if (attribute.getOperator().equals(Operator.ANY)) {
+          queryParts.add(baseSql);
+        } else {
+          if (attribute.getName().equals(NUMERICAL)) {
+            processNumericalSql(queryParts, queryParams, baseSql + AND + VALUE_AS_NUMBER, attribute);
+          } else if (attribute.getName().equals(CATEGORICAL)) {
+            processCategoricalSql(queryParts, queryParams, baseSql + AND + VALUE_AS_CONCEPT_ID, attribute);
+          } else if (attribute.getName().equals(BOTH) && attribute.getOperator().equals(Operator.IN)) {
+            processCategoricalSql(tempQueryParts, queryParams, VALUE_AS_CONCEPT_ID, attribute);
+          } else {
+            processNumericalSql(tempQueryParts, queryParams, VALUE_AS_NUMBER, attribute);
+          }
+        }
+      }
+      if (!tempQueryParts.isEmpty()) {
+        queryParts.add(baseSql + AND + "(" + String.join(OR, tempQueryParts) + ")");
+      }
+    }
+    String measurementSql = String.join(UNION_ALL, queryParts);
+    String finalSql = buildModifierSql(measurementSql, queryParams, parameters.getModifiers());
+    return QueryJobConfiguration
+      .newBuilder(finalSql)
+      .setNamedParameters(queryParams)
+      .setUseLegacySql(false)
+      .build();
+  }
+
+  @Override
+  public FactoryKey getType() {
+    return FactoryKey.MEAS;
+  }
+
+  private void validateAttributes(SearchParameter parameter) {
+    List<Attribute> attrs = parameter.getAttributes();
+    Predicate<Attribute> text = nameIsText().and(operatorAny());
+    Predicate<Attribute> numerical = nameIsNumerical().and(operatorWithCorrectOperands());
+    Predicate<Attribute> categorical = nameIsCategorical().and(inWithCorrectOperands()).or(operatorAny());
+    Predicate<Attribute> both = nameIsBoth().and(operatorWithCorrectOperands());
+    boolean textAttrs =
+      attrs.stream().filter(text::test).collect(Collectors.toList()).size() != 1;
+    boolean numericalAttrs =
+      attrs.stream().filter(numerical::test).collect(Collectors.toList()).size() != 1;
+    boolean categoricalAttrs =
+      attrs.stream().filter(categorical::test).collect(Collectors.toList()).size() != 1;
+    boolean bothAttrs =
+      attrs.stream().filter(both::test).collect(Collectors.toList()).size() == 0;
+    if (bothAttrs && textAttrs && numericalAttrs && categoricalAttrs) {
+      throw new BadRequestException("Please provide valid search attributes" +
```

Comment:
At the moment we don't really expect researchers to make their own cohort specs in code, so this isn't a huge deal... but at some point we probably will, at which point it'd be nice if we returned specific validation error messages rather than this generic message.

Reply:
I can definitely do this, but do you mind if it's in a followup PR? Will is waiting on this api call so he can complete his UI implementation of this.

Reply:
Sure, there's no rush on this. If you want to put it off until much later, fine by me, but maybe file a ticket tracking it.

Reply:
Thanks. Created a ticket for this: https://precisionmedicineinitiative.atlassian.net/browse/RW-1164